### PR TITLE
LIBCLOUD-506: Added the resource extra attribute mapping and implemented...

### PR DESCRIPTION
**Pull request by cderamus**
... it for networks, nodes and volumes. I also went ahead and added a number of additional properties for each of the three resource types.

### CODE REVIEW

File: libcloud/compute/drivers/cloudstack.py [hunk 1]
```diff
@@ -600,14 +762,21 @@ def ex_list_networks(self):
         nets = res.get('network', [])
 
         networks = []
+        extra_map = RESOURCE_EXTRA_ATTRIBUTES_MAP['network']
         for net in nets:
+            extra = self._get_extra_dict(net, extra_map)
+
+            if 'tags' in net.keys():
```
Comment (Kami) [inline libcloud/compute/drivers/cloudstack.py]: `.keys()` is unnecessary, you can just do `if 'tags' in net`.

File: libcloud/compute/drivers/cloudstack.py [hunk 2]
```diff
@@ -695,11 +864,18 @@ def list_volumes(self, node=None):
                                          method='GET')
 
         list_volumes = []
+        extra_map = RESOURCE_EXTRA_ATTRIBUTES_MAP['volume']
         for vol in volumes['volume']:
+            extra = self._get_extra_dict(vol, extra_map)
+
+            if 'tags' in vol.keys():
```
Comment (Kami) [inline libcloud/compute/drivers/cloudstack.py]: Same here.

File: libcloud/compute/drivers/cloudstack.py [hunk 3]
```diff
@@ -31,6 +31,167 @@
 from libcloud.utils.networking import is_private_subnet
 
 
+"""
+Define the extra dictionary for specific resources
+"""
+RESOURCE_EXTRA_ATTRIBUTES_MAP = {
+    'network': {
+        'broadcast_domain_type': {
+            'xpath': 'broadcastdomaintype',
```
Comment (Kami) [inline libcloud/compute/drivers/cloudstack.py]: CloudStack API returns JSON so calling a key `xpath` is deceiving :)

Maybe we should call it `key_name` or something along those lines?

Comment (cderamus) [inline libcloud/compute/drivers/cloudstack.py]: Ah, nice call!

**Comment by Kami** [comment]
Merged into trunk, thanks.

**Pull request closed without merging**